Add Cloudshell provider

roroco opened the pull request:
- Reorganize oracle providers
- Add cloudshell provider and support for using delegation token for the client

Tested with the Fn cli on a cloudshell machine.

mrc commented on provider/oracle/user_provider.go:
```diff
@@ -0,0 +1,165 @@
+	}, nil
+}
+
+//TODO support OCI environment variables
```
I wondered if this was a feature regression, but I can't see where the original code does it, so I don't think so -- let me know if I'm wrong though. If it's not removing an existing feature, Is it worth a todo? Might be better to cut an issue to track this? (Or just add support for environment variables with this PR?)

roroco replied:
Added support for OCI environment variables for the user provider with this PR.

mrc commented on provider/oracle/cloudshell_provider.go:
```diff
@@ -0,0 +1,149 @@
+	if delegationTokenFile != "" {
+		fileContent, err := ioutil.ReadFile(delegationTokenFile)
+		if err != nil {
+			return nil, fmt.Errorf("Could not load delegation token from file: %s due to: %s \n", delegationTokenFile, err)
```
Might be better not to log the filename (it could leak config into the log).
Nit: looks like an unnecessary space before the newline.

mrc commented on provider/oracle/cloudshell_provider.go:
```diff
@@ -0,0 +1,149 @@
+
+	fileContent, err := ioutil.ReadFile(DelegationTokenFileLocation)
+	if err != nil {
+		return nil, fmt.Errorf("can not load delegation_token from file %s due to: %s \n", DelegationTokenFileLocation, err)
```
Might be better not to log the filename (see line 68 comment)

mrc commented on provider/oracle/ip_provider.go:
```diff
@@ -0,0 +1,54 @@
+		if err != nil {
+			return nil, err
+		}
+		// Construct the API endpoint from the "nearby" endpoint
```
What does "nearby" mean, is it the regional endpoint? (Comment could be clearer.)

mrc commented on provider/oracle/ip_provider.go:
```diff
@@ -0,0 +1,54 @@
+		// Get the local compartment ID from the metadata endpoint
+		resp, err := http.DefaultClient.Get(CompartmentMetadata)
+		if err != nil {
+			return nil, fmt.Errorf("problem fetching compartment Id from metadata endpoint %s", err)
```
"compartment Id" -> "compartment OCID" ? (the capitalization of Id stuck out to me) (same comment on line 43)

mrc commented on provider/oracle/user_provider.go:
```diff
@@ -0,0 +1,165 @@
+
+	compartmentID := configSource.GetString(CfgCompartmentID)
+	if compartmentID == "" {
+		return nil, fmt.Errorf("no OCI compartment ID specified in config key %s ", CfgCompartmentID)
```
"compartment Id" -> "compartment OCID" ? (the capitalization of ID stuck out to me). Looks like a trailing space in the message.

mrc commented on provider/oracle/user_provider.go:
```diff
@@ -0,0 +1,165 @@
+	key, err := oci.PrivateKeyFromBytes(keyBytes, oci.String(pKeyPword))
+	if err != nil {
+		if pKeyPword != "" {
+			return nil, fmt.Errorf("Unable to load private key from file bytes: %s. Error: %s \n", pkeyFilePath, err)
```
Don't log the private key. (Nit) space before newline.

mrc commented on provider/oracle/cloudshell_provider.go:
```diff
@@ -0,0 +1,149 @@
+		csConfig.region = region
+	}
+	if csConfig.region == "" {
+		return nil, fmt.Errorf("Could not derive region from eiher config or environment.")
```
typo: eiher (same on line 61).

mrc commented on provider/oracle/cloudshell_provider.go:
```diff
@@ -0,0 +1,149 @@
+		csConfig.tenancyID = tenancyID
+	}
+	if csConfig.tenancyID == "" {
+		return nil, fmt.Errorf("Could not derive tenancy ID from eiher config or environment.")
```
typo: eiher (same on line 53). "tenancy ID" -> tenancy OCID"?

mrc commented on provider/oracle/cloudshell_provider.go:
```diff
@@ -0,0 +1,149 @@
+		csConfig.delegationToken = string(fileContent)
+	}
+	if csConfig.delegationToken == "" {
+		return nil, fmt.Errorf("Could not derive delegation token filepath from eiher config or environment.")
```
typo: eiher

mrc commented on provider/oracle/user_provider.go:
```diff
@@ -0,0 +1,165 @@
+
+	home, err := homedir.Dir()
+	if err != nil {
+		return "", nil, fmt.Errorf("error getting home directory %v", err)
```
Sometimes it's `fmt.Errorf("..%v", err)` and sometimes `fmt.Errorf("..%s", err)` (**v** or **s**) -- can you fix it up one way or t'other?

roroco commented:
> Generally good, a few nits and typos (may have been inherited from code moved around, rather than introduced by this PR) and some values we probably shouldn't be logging.

Addressed the comments.

srpurani commented:
LGTM.
